ssl/statem/statem_dtls.c

```diff
@@ -1025,6 +1078,103 @@ CON_FUNC_RETURN dtls_construct_change_cipher_spec(SSL_CONNECTION *s,
     return CON_FUNC_SUCCESS;
 }
 
+CON_FUNC_RETURN dtls_construct_ack(SSL_CONNECTION *s, WPACKET *pkt) {
+    size_t i;
+
+    if (!WPACKET_start_sub_packet_u16(pkt)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return CON_FUNC_ERROR;
+    }
+
+    for (i = 0; i < s->d1->ack_rec_num_idx; i++) {
+        /*
+         * rfc9147: section 4.
+         *
+         * Record numbers are encoded as
+         *      struct {
+         *           uint64 epoch;
+         *           uint64 sequence_number;
+         *      } RecordNumber;
+         */
+        const uint64_t epoch = s->d1->ack_rec_num[i].epoch;
+        const uint64_t sequence_number = s->d1->ack_rec_num[i].seqnum;
+
+        /*
+         * rfc9147:
+         * During the handshake, ACK records MUST be sent with an epoch which
+         * is equal to or higher than the record which is being acknowledged
+         */
+        if (epoch <= dtls1_get_epoch(s, SSL3_CC_WRITE))
+            if (!WPACKET_put_bytes_u64(pkt, epoch)
+                    || !WPACKET_put_bytes_u64(pkt, sequence_number)) {
+                SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+                return CON_FUNC_ERROR;
+            }
+    }
+
+    if (!WPACKET_close(pkt)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return CON_FUNC_ERROR;
+    }
+
+    /* Avoid acknowledging the same record numbers again */
+    s->d1->ack_rec_num_idx = 0;
+
+    return CON_FUNC_SUCCESS;
+}
+
+MSG_PROCESS_RETURN dtls_process_ack(SSL_CONNECTION *s, PACKET *pkt)
+{
+    PACKET record_numbers;
+
+    if (PACKET_remaining(pkt) == 0)
+        return MSG_PROCESS_FINISHED_READING;
+
+    if (!PACKET_get_length_prefixed_2(pkt, &record_numbers)) {
+        SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER, SSL_R_LENGTH_TOO_LONG);
+        return MSG_PROCESS_ERROR;
+    }
+
+    while (PACKET_remaining(&record_numbers) > 0) {
+        /*
+         * rfc9147: section 4.
+         *
+         * Record numbers are encoded as
+         *      struct {
+         *           uint64 epoch;
+         *           uint64 sequence_number;
+         *      } RecordNumber;
+         */
+
+        pitem *item;
+        piterator iter;
+        uint64_t epoch;
+        uint64_t sequence_number;
+
+        if (!PACKET_get_net_8(&record_numbers, &epoch)
+                || !PACKET_get_net_8(&record_numbers, &sequence_number)) {
+            SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_LENGTH_TOO_SHORT);
+            return MSG_PROCESS_ERROR;
+        }
+
+        iter = pqueue_iterator(s->d1->sent_messages);
+
+        while ((item = pqueue_next(&iter)) != NULL) {
+            dtls_sent_msg *msg = (dtls_sent_msg *)item->data;
+            size_t idx = 0;
+
+            do {
+                if (msg->rec_nums[idx].epoch == epoch
+                        && msg->rec_nums[idx].seqnum == sequence_number)
+                    msg->rec_nums[idx].acknowledged = 1;
```

Comment:
Do we need to continue searching at this point?

Reply:
I'll need to check that the we handle possible duplicate entries.